haskell-api.md

```diff
@@ -0,0 +1,236 @@
+# Haskell API Conventions
+
+## Stateless
+
+Prefer stateless endpoints. Sessions are used to authorize data access, but not
+choose what data to return.
+
+- Good:
+  - `/3/students?schools.id=x,y,z`
+  - `ApiClient.fetch("/3/students", { "schools.id": me.schoolIds })`
+- Bad:
+  - `/3/school-admins/me/students`
+  - `ApiClient.fetch("/3/school-admins/me/students")`
+
+## Parameters
+
+We want to :heart: query parameters as a way to keep the number of distinct
+endpoints lower by supporting more complex filtering combinations within each
+endpoints.
+
+- Good: `/3/students?schools.id=x,y,z&teachers.id=a,b,c`
+- Bad:
+  - `/3/schools/:id/students`
+  - `/3/teachers/:id/students`
+  - `/3/schools/:id/teachers/:id/students`
+  - etc, forever
+
+### Naming
+
+Filters should match the dotted-path of the filtered attribute in the response.
+
+For example:
+
+```
+/3/teachers
+[ { id:
+  , ...
+  , school:
+    { id:
+    , ...
+    }
+  }
+]
+```
+
+Here, a School filter would be `school.id`.
+
+```
+/3/students
+[ { id:
+  , ...
+  , schools:
+    [ { id:
+      , ...
+      }
+    ]
+  }
+]
+```
```

Comment:
I would also like to see what the request URL looks like for this response. I think it is: `/3/students?schools.id=1,2,3`.